README.md
# ember-ticketfly-accordion

_An [ARIA-compliant](https://www.w3.org/TR/wai-aria-practices-1.1/#accordion), easy-to-use
accordion system built from composable Ember components -- with
optional built-in animation through the [Web Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API)._

[![Latest NPM release][npm-badge]][npm-badge-url]
[![CircleCI Build Status][circle-badge]][circle-badge-url]
[![Test Coverage][coverage-badge]][coverage-badge-url]
[![Code Climate][codeclimate-badge]][codeclimate-badge-url]
[![Ember Observer Score][ember-observer-badge]][ember-observer-badge-url]
[![License][license-badge]][license-badge-url]
[![Dependencies][dependencies-badge]][dependencies-badge-url]
[![Dev Dependencies][devDependencies-badge]][devDependencies-badge-url]

View 

## Installing

```sh
ember install ember-ticketfly-accordion
```

## Features

- Full [WAI-ARIA](https://www.w3.org/TR/wai-aria-practices-1.1/#accordion) Compliance for Accessibility Winning.
- A robust callback API for handling DOM events with closure actions.
- Built-in (but configurable) buttery-smooth animation through the [Web-Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) and [`ember-web-animations-next-polyfill`](https://github.com/BrianSipple/ember-web-animations-next-polyfill).
- Semantic HTML
  + Tabs are comprised of `<button>` elements instead of plain `<div>`s.


## Compatibility

This addon makes use of [contextual components](http://emberjs.com/blog/2016/01/15/ember-2-3-released.html#toc_contextual-components), and is therefore
intended to support versions of Ember >= `2.3.0`.

## Usage

View the [interactive documentation](https://ticketfly.github.io/ember-ticketfly-accordion/) for usage information and tips.

## Configuration

To provide `ember-ticketfly-accordion` with [configuration options](#supported-options), define
them in a hash on the `'ember-ticketfly-accordion'` property of the object
exported from your `config/environment.js` file:

```js
ENV['ember-ticketfly-accordion'] = {
  // options go here
};
```

### Supported Options

#### Toggling Animation

Animation is enabled by default, and includes a super-lightweight implementation of
panel opening and closing built with the [Web Animations API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) &mdash; with [support back
to IE 10](https://github.com/web-animations/web-animations-js/blob/master/docs/support.md#browser-support) thanks to the
thanks to the [W3C's `web-animations-next` polyfill](https://github.com/web-animations/web-animations-next)
(enabled in Ember by [the `ember-web-animations-next-polyfill` addon](https://github.com/web-animations/web-animations-next)).

To disable animation, simply set `animatable` to `false` on the root accordion component.

To customize animation, there are two approaches:

1. Overriding the addon's default implementation by passing your own functions to
the `animatePanelOpen` and `animatePanelClosed` properties on the root accordion component (One
or both can be overridden).

2. Configuring aspects of the addon's default implementation by setting `addonAnimationSettings` properties
in `config/environment.js` like so **(available settings are shown with their current defaults)**:

```js
ENV['ember-ticketfly-accordion'] = {
  useAddonAnimations: true,
  addonAnimationSettings: {
    panelClose: {
      duration: 390
      easing: 'cubic-bezier(0.645, 0.045, 0.355, 1.000)' // ease-in-out-cubic
    },
    panelOpen: {
      duration: 390
      easing: 'cubic-bezier(0.215, 0.610, 0.355, 1)' // ease-out-cubic
    }
  }
};
```


## Working with ARIA Attributes

As part of its adherence to the [WAI-ARIA accordion spec](https://www.w3.org/TR/wai-aria-practices-1.1/#accordion), each 
part of the component system contains the `attributeBindings` needed to automatically set 
proper values for attributes such as `aria-expanded`() `aria-multiselectable`(), `aria-controls`(), etc. 

This behavior is built-in for free. 

Some ARIA attribute values are arbitrary, however, and where it's appropriate, components will declare bindings
for attributes that you can set directly. Currently, this includes:

##### `tf-accordion-panel`
- `aria-level`: Since the panel functions as a "heading", but isn't any of the standard heading elements
(`<h1>`, `<h2>`, etc), this value should be set on each panel in the accordion according to 
how you would define its heading level within your document. Since accordions can be used in just about any 
context, `tf-accordion-panel` makes no default assumptions about this in advance.



## Collaborating

* `git clone <repository-url>` this repository
* `cd ember-ticketfly-accordion`
* `npm install`
* `bower install`

### Running

* `ember serve`
* Visit your app at [http://localhost:4200](http://localhost:4200).

### Running Tests

* `npm test` (Runs `ember try:each` to test your addon against multiple Ember versions)
* `ember test`
* `ember test --server`

### Building

* `ember build`

For more information on using ember-cli, visit [https://ember-cli.com/](https://ember-cli.com/).



[npm-badge]: https://img.shields.io/npm/v/ember-ticketfly-accordion.svg
[npm-badge-url]: https://www.npmjs.com/package/ember-ticketfly-accordion
[circle-badge]: https://circleci.com/gh/Ticketfly/ember-ticketfly-accordion/tree/master.svg?style=svg&circle-token={{CIRCLE_TOKEN}}
[circle-badge-url]: https://circleci.com/gh/Ticketfly/ember-ticketfly-accordion/tree/master
[codeclimate-badge]: https://img.shields.io/codeclimate/github/Ticketfly/ember-ticketfly-accordion.svg
[codeclimate-badge-url]: https://codeclimate.com/github/Ticketfly/ember-ticketfly-accordion
[coverage-badge]: https://codeclimate.com/repos/580452d5c451cf0072003bc5/badges/fe9856d5b427c83eec3c/coverage.svg
[coverage-badge-url]: https://codeclimate.com/repos/580452d5c451cf0072003bc5/coverage
[ember-observer-badge]: http://emberobserver.com/badges/ember-ticketfly-accordion.svg
[ember-observer-badge-url]: http://emberobserver.com/addons/ember-ticketfly-accordion
[license-badge]: https://img.shields.io/npm/l/ember-ticketfly-accordion.svg
[license-badge-url]: LICENSE.md
[dependencies-badge]: https://img.shields.io/david/Ticketfly/ember-ticketfly-accordion.svg
[dependencies-badge-url]: https://david-dm.org/Ticketfly/ember-ticketfly-accordion
[devDependencies-badge]: https://img.shields.io/david/dev/Ticketfly/ember-ticketfly-accordion.svg
[devDependencies-badge-url]: https://david-dm.org/Ticketfly/ember-ticketfly-accordion#info=devDependencies


CHANGELOG.md
# 0.6.2 &mdash; February 19, 2017

- **[PATCH]**: Remove unecessary extra touch handler that was accidentally inverting panel
toggling ([#42](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/42])).


# 0.6.1 &mdash; February 17, 2017

- **[ENHANCEMENT]**: Tweak the start of the opening animation so that 
there's an extremely small duration set. ([#40](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/40])).


# 0.6.0 &mdash; February 17, 2017

- **[BREAKING CHANGE]**: Top-level accordion element no longer has `width: 100%` built-in.
It seems much better to leave `width` untouched ([#38](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/38])).

- **[ENHANCEMENT]**: Built-in styles have a little specificity as possilble, short of being behind 
the `.tfa-accordion` namespace ([#37](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/37])).

- **[ADDED]**: First "real" version of interactive documentation ([#39](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/39])).


# 0.5.0 &mdash; January 25, 2017

- **[ADDED]**: support for having a proper `aria-selected` value set on each panel tab element.
within an accordion.

- A note on the minor version bump: In retrospect, `0.4.1` also should have been a minor version bump,
since it added some significant improvements. The same applies for the addition above, thus,
here we are.


# 0.4.1 &mdash; January 25, 2017

- **[ADDED]**: `role="heading"` is now applied to the element for the `tfa-panel` component ([#32](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/32])).

- **[ADDED]**: The `tfa-panel` component now has an attribute binding for `aria-level` ([#32](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/32])).


# 0.4.0 &mdash; January 3, 2017

- **[BREAKING CHANGE]** Change name of attributes used to set class names for expanded panels ([#29](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/29])).
  + `panelExpandedTabClass` on the `tf-accordion-panel` component is now `expandedTabClassName`.
    + This syntax has a more similar feel to `expandedClassName`.
  + `panelExpandedClass` on the `tf-accordion-panel-tab` component is now `expandedClassName`.
    + This is more consistent with the `expandedClassName` property on the `tf-accordion-panel` component
    and feels more intuitive when declaring in markup.


# 0.3.1 &mdash; December 27, 2016

- **[PATCH]** Fix CHANGELOG date typo.


# 0.3.0 &mdash; December 27, 2016

- **[BREAKING CHANGE]** CSS `contain` value to `layout` and remove `overflow: hidden` 
on root accordion ([#28](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/28])).
  + `overflow` shouldn't be a concern of the addon. Simultaneously, only `layout` level
  containment is needed to optimize rendering the accordion's elements, wheras `content` containment
  prevented [`ink overflow`](https://drafts.csswg.org/css-overflow-3/#ink-overflow).

- **[ENHANCEMENT]** Add `appearance` to `none` to the styling for panel buttons ([#27](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/27])). 


# 0.2.3 &mdash; December 13, 2016

- Remove necessary (and unused) `ember-hook` dependency. ([#26](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/26]))


# 0.2.2 &mdash; December 10, 2016

- Fix `classNameBindings` to prevent `undefined` from being added. ([#25](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/25]))


# 0.2.1 &mdash; December 9, 2016

- Bind the "tfa-panel--expanded" class to panel elements ([#22](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/22]))

- Unset browsers's override of the CSS `text-align` property 
  for the panel tab (button) element ([#23](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/23]))

- Clean up handling of custom class names. Support custom expanded-state classes ([#24](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/24]))


# 0.2.0 &mdash; December 8, 2016

- Officially nix the idea for configurable stylesheets ([#20](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/20]))


# 0.1.0 &mdash; December 1, 2016

- Enable `multi-expand mode` ([#12](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/12))

- Implement, document, and test configurable animations ([#13](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/13))

- Implement, document, and test configurable animation lifecycle hooks ([#14](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/14))

- Implement, document, and test interface for events and action handling ([#18](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/18))


# 0.0.3 &mdash; November 5, 2016

- Move styles to `vendor/` directory and import then
selectively from there ([#11](https://github.com/Ticketfly/ember-ticketfly-accordion/pull/11).)


# 0.0.2 &mdash; November 5, 2016

- Add CSS layout containment to root accordion. 


# 0.0.1 &mdash; November 5, 2016

- Initial core functionality and testing for accordion behavior


addon/components/tf-accordion.js
import Component from 'ember-component';
import layout from '../templates/components/tf-accordion';
import get from 'ember-metal/get';
import set from 'ember-metal/set';
import computed, { notEmpty, bool } from 'ember-computed';
import { A } from 'ember-array/utils';
import { scheduleOnce } from 'ember-runloop';
import { animatePanelOpen, animatePanelClosed } from 'ember-ticketfly-accordion/utils/accordion-panel-animation';

/**
 * @module tf-accordion
 */

/**
 * @class TFAccordionComponent
 * @namespace TFAccordion
 * @extends Ember.Component
 */
export default Component.extend({
  layout,
  classNames: ['tfa-accordion'],
  attributeBindings: ['aria-multiselectable'],
  ariaRole: 'tablist',
  'aria-multiselectable': 'true', // @see {@link https://github.com/BrianSipple/why-am-i-doing-this/blob/master/ember/aria-attribute-binding-in-components.md}

  handleKeydown: null, // TOOD: Rename to 'on-keydown'

  KEYCODE_LEFT: 37,
  KEYCODE_UP: 38,
  KEYCODE_RIGHT: 39,
  KEYCODE_DOWN: 40,

  focusedIndex: -1,

  /* ---------- API ---------- */

  /**
   * Whether or not all panels can be expanded at once.
   *
   * Default "accordion" behavior consists of collapsing all
   * but the most-recently expanded panel
   */
  multiExpand: false,

  /**
   * Whether or not focus should cycle around the panel during
   * arrow-key navigation.
   *
   * For example, when the last panel is focused and the down arrow
   * is pressed, cycling focus will focus the first panel. Otherwise, the
   * last panel will remain focused.
   */
  cycleFocus: true,

  /**
   * Whether or not animation is enabled for expanding/collapsing a panel
   */
  animatable: true,

  /**
   * function to execute when animating the panel closed
   */
  animatePanelClosed: null,

  /**
   * function to execute when animating the panel closed
   */
  animatePanelOpen: null,


  /**
   * If the default opening animation is used, this callback
   * will be called when it completes -- passing in the component
   * object of the panel that was animated.
   */
  onPanelAnimatedOpen: null,

  /**
   * If the default closing animation is used, this callback
   * will be executed when it completes -- passing in the component
   * object of the panel that was animated.
   */
  onPanelAnimatedClosed: null,


  panels: computed(function() {
    return A();
  }),

  /* ---------- COMPUTEDS ---------- */

  isAnimatable: bool('animatable'),
  hasPanels: notEmpty('panels'),

  /* ---------- LIFECYCLE ---------- */

  init() {
    this._super(...arguments);

    this._setAnimationFunctionsOnInit();
  },

  willDestroyElement() {
    this._super(...arguments);

    scheduleOnce('actions', this, function () {
      set(this, 'panels', null);
    });
  },

  /* ---------- COMPONENT ELEMENT EVENT LISTENERS ---------- */

  keyDown(ev) {
    const keyCode = ev.keyCode || ev.which;

    if (keyCode === this.KEYCODE_LEFT || keyCode === this.KEYCODE_UP) {
      ev.preventDefault();
      this.incrementPanelFocus(-1);
    }

    if (keyCode === this.KEYCODE_RIGHT || keyCode === this.KEYCODE_DOWN) {
      ev.preventDefault();
      this.incrementPanelFocus(1);
    }
  },

  /* ---------- PUBLIC METHODS ---------- */

  incrementPanelFocus(increment) {
    const cycleFocus = get(this, 'cycleFocus');
    const focusedIndex = get(this, 'focusedIndex');
    const numPanels = get(this, 'panels.length');
    const { max, min } = Math;

    let nextIndex;

    if (cycleFocus) {
      nextIndex = (focusedIndex + increment + numPanels) % numPanels;

    } else {
      nextIndex = increment > 0 ?
        min(numPanels - 1, focusedIndex + 1)
        :
        max(0, focusedIndex -1);
    }

    const panelToFocus = get(this, 'panels').objectAt(nextIndex);

    scheduleOnce('afterRender', this, 'setFocusOnPanel', panelToFocus, nextIndex);
  },

  setFocusOnPanel(panel, index) {
    set(this, 'focusedIndex', index);
    get(this, 'panels').setEach('isFocused', false);
    set(panel, 'isFocused', true);
    get(panel, 'tab.element').focus();
  },

  /**
   * Adds a panel from the `panels` array
   *
   * @method registerPanel
   * @param {TFAccordion.TFAccordionPanel Component} panel
   * @public
   */
  registerPanel(panel) {
    get(this, 'panels').addObject(panel);
  },

  /**
   * Removes a panel from the `panels` array
   *
   * @method unRegisterPanel
   * @param {TFAccordion.TFAccordionPanel Component} panel
   * @public
   */
  unRegisterPanel(panel) {
    get(this, 'panels').removeObject(panel);
  },

  /* ---------- ACTIONS ---------- */

  actions: {
    onPanelSelection(panel) {
      const panels = get(this, 'panels');
      const indexOfSelected = panels.indexOf(panel);
      const shouldExpand = !get(panel, 'isExpanded');
      const isAnimatable = get(this, 'isAnimatable');

      this._handleMultiExpandOnPanelSelect(indexOfSelected, panels, isAnimatable);

      scheduleOnce('afterRender', this, 'setFocusOnPanel', panel, indexOfSelected);

      if (isAnimatable && !get(panel, 'isInMotion')) {
        const animationFunc = shouldExpand ? 'animatePanelOpen' : 'animatePanelClosed';
        const animationCompleteCallback = shouldExpand ? 'onPanelAnimatedOpen' : 'onPanelAnimatedClosed';

        set(panel, 'isInMotion', true);
        scheduleOnce('afterRender', this, animationFunc, panel, this[animationCompleteCallback]);

      } else {
        scheduleOnce('afterRender', this, function updatePanelExpansionState() {
          set(panel, 'isExpanded', shouldExpand);
        });
      }
    }
  },

  /* ---------- PRIVATE HELPER METHODS ---------- */

  _onPanelAnimatedOpen(panelComponent) {
    set(panelComponent, 'isInMotion', false);
  },

  _onPanelAnimatedClosed(panelComponent) {
    set(panelComponent, 'isInMotion', false);
  },

  /**
   * If we're not in multi-expand mode, and a different panel was
   * selected than the one currently open, we need to close it
   */
  _handleMultiExpandOnPanelSelect(indexOfSelected, panels, useAnimation) {
    const currentlyFocusedIndex = get(this, 'focusedIndex');
    const multiExpand = get(this, 'multiExpand');

    if (
      !multiExpand &&
      (
        currentlyFocusedIndex >= 0 &&
        indexOfSelected !== currentlyFocusedIndex
      )
    ) {
      const currentlyExpandedPanel = panels.objectAt(currentlyFocusedIndex);

      if (useAnimation) {
        scheduleOnce('afterRender', this, 'animatePanelClosed', currentlyExpandedPanel);

      } else {
        scheduleOnce('afterRender', this, function updatePanelExpansionState() {
          set(currentlyExpandedPanel, 'isExpanded', false);
        });
      }
    }
  },

  /**
   * If no animation function is provided for opening/closing panel body elements,
   * we default to our utility
   */
  _setAnimationFunctionsOnInit() {
    this.animatePanelOpen = (typeof this.animatePanelOpen === 'function') ? this.animatePanelOpen : animatePanelOpen;
    this.animatePanelClosed = (typeof this.animatePanelClosed === 'function') ? this.animatePanelClosed : animatePanelClosed;

    this.onPanelAnimatedOpen = (typeof this.onPanelAnimatedOpen === 'function') ? this.onPanelAnimatedOpen : this._onPanelAnimatedOpen;
    this.onPanelAnimatedClosed = (typeof this.onPanelAnimatedClosed === 'function') ? this.onPanelAnimatedClosed : this._onPanelAnimatedClosed;
  }
});
